wp_api/src/auth.rs

```diff
@@ -0,0 +1,138 @@
+use http::HeaderValue;
+use std::sync::{Arc, RwLock};
+
+#[derive(Debug, Clone, uniffi::Enum)]
+pub enum WpAuthentication {
+    AuthorizationHeader { token: String },
+    Bearer { token: String },
+    None,
+}
+
+#[uniffi::export]
+fn wp_authentication_from_username_and_password(
+    username: String,
+    password: String,
+) -> WpAuthentication {
+    WpAuthentication::from_username_and_password(username, password)
+}
+
+impl WpAuthentication {
+    pub fn from_username_and_password(username: String, password: String) -> Self {
+        use base64::prelude::*;
+        WpAuthentication::AuthorizationHeader {
+            token: BASE64_STANDARD.encode(format!("{}:{}", username, password)),
+        }
+    }
+
+    pub fn header_value(&self) -> Option<HeaderValue> {
+        match self {
+            Self::None => None,
+            Self::AuthorizationHeader { token } => {
+                Some(HeaderValue::from_str(&format!("Basic {}", token))
+                .expect("It shouldn't be possible to build WpAuthentication::AuthorizationHeader with an invalid token"))
+            }
+            Self::Bearer { token } => {
+                Some(HeaderValue::from_str(&format!("Bearer {}", token)).expect("It shouldn't be possible to build WpAuthentication::Bearer with an invalid token"))
+            }
+        }
+    }
+}
+
+#[derive(uniffi::Object)]
+pub struct ModifiableAuthenticationProvider {
+    auth: RwLock<WpAuthentication>,
+}
+
+#[uniffi::export]
+impl ModifiableAuthenticationProvider {
+    #[uniffi::constructor]
+    pub fn new(authentication: WpAuthentication) -> Self {
+        Self {
+            auth: RwLock::new(authentication),
+        }
+    }
+
+    pub fn set_authentication(&self, new_authentication: WpAuthentication) {
+        *self
+            .auth
+            .write()
+            .expect("If the lock is poisoned, there isn't much we can do") = new_authentication;
+    }
+}
+
+impl ModifiableAuthenticationProvider {
+    pub fn header_value(&self) -> Option<HeaderValue> {
+        self.auth
+            .read()
+            .expect("If the lock is poisoned, there isn't much we can do")
+            .header_value()
+    }
+}
+
+#[uniffi::export(with_foreign)]
+pub trait WpDynamicAuthenticationProvider: Send + Sync {
+    fn auth(&self) -> WpAuthentication;
+}
+
+#[derive(uniffi::Object)]
+pub enum WpAuthenticationProvider {
+    StaticAuthenticationProvider {
+        auth: WpAuthentication,
+    },
+    DynamicAuthenticationProvider {
+        inner: Arc<dyn WpDynamicAuthenticationProvider>,
+    },
+    Modifiable {
+        inner: Arc<ModifiableAuthenticationProvider>,
+    },
+}
```

Comment:
I'm not convinced we need static/dynamic/modifiable?

The app layer should be able to handle this – if we want to fetch from some backing store constantly, that's fine. If we want to just keep a value in memory and have it returned for every request, it can do that too.

But having the complexity here IMHO isn't super valuable?

Reply:
We need at least two of the three and having the third doesn't add any extra complexity. In fact, this setup is fairly straightforward, so I don't think we need to be worried about it.

Reply:
> We need at least two of the three 

Sorry, why is this? IINM `DynamicAuthenticationProvider` can do anything the other two can, no?

Reply:
I believe this is covered in the PR description, but if that's not clear, happy to discuss it further.